curl/src/main/scala/org/http4s/curl/http/CurlRequest.scala

```diff
@@ -0,0 +1,92 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.curl.http
+
+import cats.effect._
+import org.http4s.Request
+import org.http4s.Response
+import org.http4s.curl.internal.Utils
+import org.http4s.curl.internal._
+import org.http4s.curl.unsafe.CurlExecutorScheduler
+
+private[curl] object CurlRequest {
+  private def setup(
+      handle: CurlEasy,
+      ec: CurlExecutorScheduler,
+      send: RequestSend,
+      recv: RequestRecv,
+      req: Request[IO],
+  ): Resource[IO, Unit] = Utils.newZone.flatMap(implicit zone =>
+    CurlSList().evalMap(headers =>
+      IO {
+        // TODO add in options
+        // handle.setVerbose(true)
+
+        import org.http4s.curl.unsafe.libcurl_const
+        import scala.scalanative.unsafe._
+        import org.http4s.Header
+        import org.http4s.HttpVersion
+        import org.typelevel.ci._
+
+        handle.setCustomRequest(toCString(req.method.renderString))
+
+        handle.setUpload(true)
+
+        handle.setUrl(toCString(req.uri.renderString))
+
+        val httpVersion = req.httpVersion match {
+          case HttpVersion.`HTTP/1.0` => libcurl_const.CURL_HTTP_VERSION_1_0
+          case HttpVersion.`HTTP/1.1` => libcurl_const.CURL_HTTP_VERSION_1_1
+          case HttpVersion.`HTTP/2` => libcurl_const.CURL_HTTP_VERSION_2
+          case HttpVersion.`HTTP/3` => libcurl_const.CURL_HTTP_VERSION_3
+          case _ => libcurl_const.CURL_HTTP_VERSION_NONE
+        }
+        handle.setHttpVersion(httpVersion)
+
+        req.headers // curl adds these headers automatically, so we explicitly disable them
+          .transform(Header.Raw(ci"Expect", "") :: Header.Raw(ci"Transfer-Encoding", "") :: _)
+          .foreach(header => headers.append(header.toString))
+
+        handle.setHttpHeader(headers.toPtr)
+
+        // NOTE that pointers are to data that is handled by GC
+        // and have proper lifetime management
+        // so no need to handle them explicitly anymore
+        handle.setReadData(Utils.toPtr(send))
+        handle.setReadFunction(RequestSend.readCallback(_, _, _, _))
+
+        handle.setHeaderData(Utils.toPtr(recv))
+        handle.setHeaderFunction(RequestRecv.headerCallback(_, _, _, _))
+
+        handle.setWriteData(Utils.toPtr(recv))
+        handle.setWriteFunction(RequestRecv.writeCallback(_, _, _, _))
+
+        ec.addHandle(handle.curl, recv.onTerminated)
+      }
+    )
+  )
+
+  def apply(ec: CurlExecutorScheduler, req: Request[IO]): Resource[IO, Response[IO]] = for {
+    handle <- CurlEasy()
+    flow <- FlowControl(handle)
+    send <- RequestSend(flow)
+    recv <- RequestRecv(flow)
```

Comment:
This `send` and `recv` are the ones whose pointers are passed to cURL, correct?

These are both `Resource`s where the `RequestSend` and `RequestRecv` objects are created via a `.map` I believe, so there will be no references kept to them there after the map executes. So the references must be kept somewhere else. Somewhere in the next three lines.

But where exactly? It would be good if this was made explicit, so it doesn't accidentally get refactored away for example.

Reply:
Oh, I see the problem now 👍; those objects are not used directly anymore after setup, so there's nothing that prevents GC from cleaning them up.
I think I need to check websocket handler too, as it is also something like this. I'll update the PR with fixed version tomorrow.